src/gensxg.c

```diff
@@ -0,0 +1,334 @@
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <openssl/pem.h>
+
+#include "libsxg.h"
+
+const struct option longopts[] = {
+  { "help",         no_argument,       NULL, 'h' },
+  { "url",          required_argument, NULL, 'u' },
+  { "certUrl",      required_argument, NULL, 'r' },
+  { "validityUrl",  required_argument, NULL, 'v' },
+  { "certificate",  required_argument, NULL, 'e' },
+  { "publicKey",    required_argument, NULL, 'p' },
+  { "privateKey",   required_argument, NULL, 'k' },
+  { "content",      required_argument, NULL, 'c' },
+  { "contentType",  required_argument, NULL, 't' },
+  { "date",         required_argument, NULL, 'd' },
+  { "expire",       required_argument, NULL, 'x' },
+  { "miRecordSize", required_argument, NULL, 'm' },
+  { 0,              0,                 0,     0  },
+};
+
+const char kHelpMessage[] =
+    "-help\n"
+    "  Show this message.\n"
+    "-url string\n"
+    "  The URI of the resource represented in the SXG file. (required)\n"
+    "-certUrl string\n"
+    "  The URI of certificate cbor file published. (required)\n"
+    "-validityUrl string\n"
+    "  The URI of validity information provided. (required)\n"
+    "-certificate string\n"
+    "  The certificate PEM file for the SXG. (mutual exclusive with "
+    "-publicKey)\n"
+    "-publicKey string\n"
+    "  The Ed25519 PEM file for the SXG. (mutual exclusive with "
+    "-certificate)\n"
+    "-private_key string\n"
+    "  The private key PEM file for the SXG. (required)\n"
+    "-content string\n"
+    "  Source file to be used as SXG payload (default [index.html]).\n"
+    "-contentType string\n"
+    "  Mime type of Source file (default [text/html]).\n"
+    "-date int\n"
+    "  The datetime for the SXG in unixtime. Use now by default.\n"
+    "-expire int\n"
+    "  The expire time of the SXG in unixtime. (default <date> +7 days)\n"
+    "-miRecordSize int\n"
+    "  The record size of Merkle Integrity Content Encoding. "
+    "(default [4096])\n";
+
+typedef struct {
+  bool help;
+  const char* url;
+  const char* cert_url;
+  const char* validity_url;
+  const char* certificate;
+  const char* public_key;
+  const char* private_key;
+  const char* content;
+  const char* content_type;
+  uint64_t date;
+  uint64_t expire;
+  uint64_t mi_record_size;
+} Options;
+
+static EVP_PKEY* load_private_key(const char* filepath) {
+  FILE* const keyfile = fopen(filepath, "r");
+  if (keyfile == NULL) {
+    return NULL;
+  }
+  EVP_PKEY* private_key =
+      PEM_read_PrivateKey(keyfile, NULL, NULL, NULL);
+  fclose(keyfile);
+  return private_key;
+}
+
+static X509* load_x509_cert(const char* filepath) {
+  FILE* certfile = fopen(filepath, "r");
+  if (certfile == NULL) {
+    return NULL;
+  }
+  X509* cert = PEM_read_X509(certfile, NULL, NULL, NULL);
+  fclose(certfile);
+  return cert;
+}
+
+static EVP_PKEY* load_ed25519_pubkey(const char* filepath) {
+  FILE* keyfile = fopen(filepath, "r");
+  if (!keyfile) {
```

Comment:
`keyfile != NULL`? You write `SOME_PTR == NULL` several times so better to be consistent.

Reply:
done.